src/backend/mps/compute.rs

```diff
@@ -18,23 +23,408 @@ impl MpsCompute {
+        k: usize,
+    ) -> Result<Buffer, MpsError> {
+        if m == 0 || n == 0 || k == 0 {
+            return Err(MpsError::InvalidDimensions);
+        }
+
+        let result_size = m * k * std::mem::size_of::<f32>();
+        let result_buffer = self
+            .device
+            .device()
+            .new_buffer(result_size as u64, MTLResourceOptions::StorageModeShared);
+
+        // Create dimension buffers
+        let m_buffer = self.create_buffer(&[m as u32])?;
+        let n_buffer = self.create_buffer(&[n as u32])?;
+        let k_buffer = self.create_buffer(&[k as u32])?;
+
+        let library = self
+            .device
+            .device()
+            .new_library_with_source(
+                include_str!("../../../shaders/metal/matrix_ops.metal"),
+                &metal::CompileOptions::new(),
+            )
+            .map_err(|_| MpsError::ShaderCompilationError)?;
+
+        let kernel = library
+            .get_function("matrix_multiply", None)
+            .map_err(|_| MpsError::ShaderCompilationError)?;
+
+        let pipeline = self
+            .device
+            .device()
+            .new_compute_pipeline_state_with_function(&kernel)
+            .map_err(|_| MpsError::ShaderCompilationError)?;
+
+        let command_buffer = self.command_queue.new_command_buffer();
+        let compute_encoder = command_buffer.new_compute_command_encoder();
+
+        compute_encoder.set_compute_pipeline_state(&pipeline);
+        compute_encoder.set_buffer(0, Some(a), 0);
+        compute_encoder.set_buffer(1, Some(b), 0);
+        compute_encoder.set_buffer(2, Some(&result_buffer), 0);
+        compute_encoder.set_buffer(3, Some(&m_buffer), 0);
+        compute_encoder.set_buffer(4, Some(&n_buffer), 0);
+        compute_encoder.set_buffer(5, Some(&k_buffer), 0);
+
+        let num_threads = pipeline.thread_execution_width();
+
+        // Create thread groups
+        let thread_group_count = MTLSize {
+            width: (((m * k) as NSUInteger + num_threads) / num_threads),
+            height: 1,
+            depth: 1,
+        };
+        let thread_group_size = MTLSize {
+            width: num_threads,
+            height: 1,
+            depth: 1,
+        };
+
+        compute_encoder.dispatch_thread_groups(thread_group_count, thread_group_size);
+        compute_encoder.end_encoding();
+
+        command_buffer.commit();
+        command_buffer.wait_until_completed();
+
+        Ok(result_buffer)
+    }
+
+    pub fn add(&self, a: &Buffer, b: &Buffer, size: usize) -> Result<Buffer, MpsError> {
+        let result_buffer = self.device.device().new_buffer(
+            (size * size_of::<f32>()) as u64,
+            MTLResourceOptions::StorageModeShared,
+        );
+
+        // Create and compile the addition kernel
+        let library = self
+            .device
+            .device()
+            .new_library_with_source(
+                include_str!("../../../shaders/metal/binary_ops.metal"),
```

Comment:
Is there a way to transfer this section to build.rs too? or set the library from the `new()` and call it from `self`

> Me being Lazy to point out the lines from the start, commenting on every single line XD